src/pygama/raw/orca/orca_digitizers.py

```diff
@@ -210,3 +204,228 @@ def decode_packet(self, packet, packet_id, rbl):
         return evt_rbkd[ccc].is_full()
+
+
+class ORSIS3316WaveformDecoder(OrcaDecoder):
+    """Decoder for `Struck SIS3316 <https://www.struck.de/sis3316.html>`_
+    digitizer data written by ORCA."""
+
+    def __init__(self, header: OrcaHeader = None, **kwargs) -> None:
+
+        # store an entry for every event
+        self.decoded_values_template = {
+            "packet_id": {
+                "dtype": "uint32",
+            },
+            "energy": {
+                "dtype": "uint32",
+                "units": "adc",
+            },
+            "energy_first": {
+                "dtype": "uint32",
+            },
+            "timestamp": {
+                "dtype": "uint64",
+                "units": "clock_ticks",
+            },
+            "crate": {
+                "dtype": "uint8",
+            },
+            "card": {
+                "dtype": "uint8",
+            },
+            "channel": {
+                "dtype": "uint8",
+            },
+            # waveform data
+            "waveform": {
+                "dtype": "uint16",
+                "datatype": "waveform",
+                "wf_len": 65532,  # max value. override this before initializing buffers to save RAM
+                "dt": 8,  # override if a different clock rate is used
+                "dt_units": "ns",
+                "t0_units": "ns",
+            },
+        }
+
+        self.decoded_values = {}
+        self.skipped_channels = {}
+        super().__init__(
+            header=header, **kwargs
+        )  # also initializes the garbage df (whatever that means...)
+
+    def set_header(self, header: OrcaHeader) -> None:
+        self.header = header
+
+        self.decoded_values = copy.deepcopy(self.decoded_values_template)
+
+        for card_dict in self.header["ObjectInfo"]["Crates"][0]["Cards"]:
+            if card_dict["Class Name"] == "ORSIS3316Model":
+                card = card_dict["Card"]
+                crate = 0
+                for channel in range(0, 16):
+                    ccc = get_ccc(crate, card, channel)
+                    trace_length = card_dict["rawDataBufferLen"]
+                    self.decoded_values[ccc] = copy.deepcopy(
+                        self.decoded_values_template
+                    )
+
+                    if trace_length <= 0 or trace_length > 2**16:
+                        raise RuntimeError(
+                            "invalid trace_length: ",
+                            trace_length,
+                        )
+
+                    self.decoded_values[ccc]["waveform"]["wf_len"] = trace_length
+
+        self.decoded_values["waveform"]["wf_len"] = trace_length
+
+    def get_key_list(self) -> list[int]:
+        key_list = []
+        for key in self.decoded_values.keys():
+            key_list += [key]
+        return key_list
+
+    def get_decoded_values(self, key: int = None) -> dict[str, Any]:
+        if key is None:
+            dec_vals_list = self.decoded_values
```

Comment:
note: dec_vals_list is a dict, not a list...

I suggest setting it to self.decoded_values.values()